[UPDATED] Logging based on changes to gnatsd logger

kozlovic opened the pull request:
Logging in NATS Streaming has been reworked.

tylertreat commented on logger/logger.go:
```diff
@@ -0,0 +1,97 @@
+	}
+	f(s.log, LogPrefix+format, args...)
+	s.mu.Unlock()
+}
```
This code looks pretty similar to the gnatsd logger. I think at some point we should think about making a NATS logger library.

tylertreat commented on server/server.go:
```diff
@@ -688,30 +703,26 @@ var DefaultNatsServerOptions = server.Options{
-func stanClosedHandler(nc *nats.Conn) {
-	Debugf("connection %q has been closed", nc.Opts.Name)
+func (s *StanServer) stanClosedHandler(nc *nats.Conn) {
+	if s.debug {
```
Why the check? Logger should handle the conditional logging, right?

tylertreat commented on server/server.go:
```diff
@@ -821,7 +832,9 @@ func (s *StanServer) createNatsClientConn(name string, sOpts *Options, nOpts *se
 	}
 
-	Tracef(" NATS conn opts: %v", ncOpts)
+	if s.trace {
```
Same question as above.

kozlovic replied:
For performance. Early on, I realized that the cost in calling Tracef or Debugf is substantial in term of string allocation (if there is formatting involved, etc..), so it was better to do the check at the higher layer. I have kept the check at the lower level in case one forgets to do the check before calling those functions, and to respect the contract of the logger.

tylertreat replied:
Should we make note of this in a comment? It might not be obvious to future maintainers.

kozlovic replied:
But there are many places in the code where I do that. It happens that you saw this one because of the way the code review is.

kozlovic replied:
If you want to check what the server's `debug` and `trace` flags are, you would find this in the `StanServer` structure. I think that's enough:
```
	// Use these flags for Debug/Trace in places where speed matters.
	// Normally, Debugf and Tracef will check an atomic variable to
	// figure out if the statement should be logged, however, the
	// cost of calling Debugf/Tracef is still significant since there
	// may be memory allocations to format the string passed to these
	// calls. So in those situations, use these flags to surround the
	// calls to Debugf/Tracef.
```
I will remove the `atomic` reference. In this PR, I also surrounded the Debug print statements about subjects with `if s.debug`, but that was unnecessary. As the comment suggest, you would want to do that only when speed matters (that is, in the fast path). I am going to look at the place where I use the server's debug/trace and remove it in place where it is not really needed.

kozlovic commented on logger/logger_test.go:
```diff
@@ -0,0 +1,117 @@
+// Copyright 2016 Apcera Inc. All rights reserved.
```
Will fix the copyright year here

ColinSullivan1 commented:
@kozlovic , was there an issue, or is this maintenance?

ColinSullivan1 commented on stores/limits.go:
```diff
@@ -19,16 +19,23 @@ const (
 	return &cloned
 }
 
+// ClonePerChannelMap returns a deep copy of the StoreLimits's PerChannel map
```
IMO this should be in a different PR; it's not logging related.

kozlovic commented:
@tylertreat @ColinSullivan1 I have removed the checks to s.debug/trace where not needed and removed the Clone() usage for this PR (once this is merged, I will submit a new one).
Let me know if this is ok for merge. Thanks!

tylertreat commented:
lgtm
